elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +266,382 @@ private Object fetchRelationship(Environment context) {
+         */
+        private Triplet(List<Map<String, Object>> input, PersistentResource parentResource, String fieldName) {
+            this.input = input;
+            this.parentResource = parentResource;
+            this.fieldName = fieldName;
+        }
+    }
+
+    /**
+     * handles both updating/creating by first handling the attributes and then the relationships, queuing up
+     * any further nested relationships
+     * @param input data input
+     * @param parent parent
+     * @param requestScope request scope
+     * @param fieldName field name
+     * @return set of persistent resource objects
+     */
+    private Set<PersistentResource> inputObjectToUpsertOrUpdateOperations(Map<String, Object> input, Optional<String> id,
+                                                                          PersistentResource parent, RequestScope requestScope,
+                                                                          String fieldName, GraphQLType outputType) {
+        EntityDictionary dictionary = requestScope.getDictionary();
+        Set<PersistentResource> returnOps = new HashSet();
+        Set<Triplet> visited = new HashSet();
+        Queue<Triplet> toVisit = new LinkedList();
+        toVisit.add(new Triplet(new ArrayList(Arrays.asList(input)), parent, fieldName));
+
+        /* loop through the queue */
+        while(!toVisit.isEmpty()) {
+            /* pick the head of the queue to process */
+            Triplet triplet = toVisit.poll();
+            Class<?> entityClass;
+
+            /* avoid cycles */
+            if(visited.contains(triplet)) {
+                continue;
+            } else {
+                visited.add(triplet);
+            }
+
+            /* get entity class of either root or relationship (in case of non-root) */
+            if(triplet.parentResource == null) {
+                entityClass = dictionary.getEntityClass(triplet.fieldName);
+            } else {
+                entityClass = dictionary.getParameterizedType(triplet.parentResource.getResourceClass(), triplet.fieldName);
+            }
+
+            /* fetch just the attributes from current input object */
+            Map<String, Object> attributesOnly = stripRelationships(triplet.input, entityClass, dictionary);
+
+            /* fetch ALL the relationships from current input object */
+            Map<String, Object> relationshipsOnly = stripAttributes(triplet.input, entityClass, dictionary);
+
+            /* new parent will be the resource we create from the stripped attributes */
+            PersistentResource newParent = updateOrCreateAttributes(attributesOnly, id, requestScope, entityClass,
+                    outputType, triplet.fieldName, triplet.parentResource);
+
+            returnOps.add(newParent);
+
+            /* loop through each relation */
+            for(Map.Entry<String, Object> relationship : relationshipsOnly.entrySet()) {
+                Object value = relationship.getValue();
```

Comment:
This function can be greatly simplified by simply adding the relationship entity to the toVisit list.  You don't need to do anything else inside this loop.